refactor: Refactor market related code

sl1depengwyn opened the pull request:
Closes #10867
Closes #10512
Closes https://github.com/blockscout/blockscout/pull/10320

## Changelog

- Single behaviour for market data provider (source): `Explorer.Market.Source`
- Three source agnostic fetchers put in the folder `Explorer.Market.Fetcher` and uses only behaviour callbacks
- Coin id is required for native and secondary coin fetching and history to avoid fetching of wrong coin by `COIN` env var
- Tokens exchange rates enabled by default, but won't work until any source is configured for tokens fetching (i.e. `EXCHANGE_RATES_COINGECKO_PLATFORM_ID` for CoinGecko), so now one variable is needed to enable tokens market data fetching
- `tokens.icon_url` type changed from `varchar(255)` to `text`

### Env vars changes:

#### Removed:
- `EXCHANGE_RATES_CRYPTORANK_LIMIT` This is part of fetcher config, not source
- `EXCHANGE_RATES_COIN` Not used anymore

#### Added:
- `MARKET_SECONDARY_COIN_HISTORY_SOURCE`
- `MARKET_COIN_FETCHER_ENABLED`
- `MARKET_HISTORY_FETCHER_ENABLED`
- `MARKET_CRYPTOCOMPARE_BASE_URL`
- `MARKET_DEFILLAMA_COIN_ID`
- `MARKET_CRYPTOCOMPARE_COIN_SYMBOL`
- `MARKET_MOBULA_BASE_URL`
- `MARKET_HISTORY_FIRST_FETCH_DAY_COUNT`

#### Renamed:
- `EXCHANGE_RATES_SOURCE` -> `MARKET_NATIVE_COIN_SOURCE`
- `TOKEN_EXCHANGE_RATES_SOURCE` -> `MARKET_TOKENS_SOURCE`
- `EXCHANGE_RATES_PRICE_SOURCE` -> `MARKET_NATIVE_COIN_HISTORY_SOURCE`
- `EXCHANGE_RATES_MARKET_CAP_SOURCE` -> `MARKET_MARKET_CAP_HISTORY_SOURCE`
- `EXCHANGE_RATES_TVL_SOURCE` -> `MARKET_TVL_HISTORY_SOURCE`
- `DISABLE_TOKEN_EXCHANGE_RATE` -> `MARKET_TOKENS_FETCHER_ENABLED`
- `TOKEN_EXCHANGE_RATE_INTERVAL` -> `MARKET_TOKENS_INTERVAL`
- `TOKEN_EXCHANGE_RATE_REFETCH_INTERVAL` -> `MARKET_TOKENS_REFETCH_INTERVAL`
- `TOKEN_EXCHANGE_RATE_MAX_BATCH_SIZE` -> `MARKET_TOKENS_MAX_BATCH_SIZE`
- `CACHE_EXCHANGE_RATES_PERIOD` -> `MARKET_COIN_CACHE_PERIOD`
- `DISABLE_EXCHANGE_RATES` -> `DISABLE_MARKET`
- `EXCHANGE_RATES_SECONDARY_COIN_SOURCE` -> `MARKET_SECONDARY_COIN_SOURCE`
- `EXCHANGE_RATES_COINGECKO_PLATFORM_ID` -> `MARKET_COINGECKO_PLATFORM_ID`
- `EXCHANGE_RATES_COINGECKO_BASE_URL` -> `MARKET_COINGECKO_BASE_URL`
- `EXCHANGE_RATES_COINGECKO_API_KEY` -> `MARKET_COINGECKO_API_KEY`
- `EXCHANGE_RATES_COINGECKO_COIN_ID` -> `MARKET_COINGECKO_COIN_ID`
- `EXCHANGE_RATES_COINGECKO_SECONDARY_COIN_ID` -> `MARKET_COINGECKO_SECONDARY_COIN_ID`
- `EXCHANGE_RATES_COINMARKETCAP_BASE_URL` -> `MARKET_COINMARKETCAP_BASE_URL`
- `EXCHANGE_RATES_COINMARKETCAP_API_KEY` -> `MARKET_COINMARKETCAP_API_KEY`
- `EXCHANGE_RATES_COINMARKETCAP_COIN_ID` -> `MARKET_COINMARKETCAP_COIN_ID`
- `EXCHANGE_RATES_COINMARKETCAP_SECONDARY_COIN_ID` -> `MARKET_COINMARKETCAP_SECONDARY_COIN_ID`
- `EXCHANGE_RATES_CRYPTOCOMPARE_SECONDARY_COIN_SYMBOL` -> `MARKET_CRYPTOCOMPARE_SECONDARY_COIN_SYMBOL`
- `EXCHANGE_RATES_CRYPTORANK_PLATFORM_ID` -> `MARKET_CRYPTORANK_PLATFORM_ID`
- `EXCHANGE_RATES_CRYPTORANK_BASE_URL` -> `MARKET_CRYPTORANK_BASE_URL`
- `EXCHANGE_RATES_CRYPTORANK_API_KEY` -> `MARKET_CRYPTORANK_API_KEY`
- `EXCHANGE_RATES_CRYPTORANK_COIN_ID` -> `MARKET_CRYPTORANK_COIN_ID`
- `EXCHANGE_RATES_CRYPTORANK_SECONDARY_COIN_ID` -> `MARKET_CRYPTORANK_SECONDARY_COIN_ID`
- `EXCHANGE_RATES_MOBULA_CHAIN_ID` -> `MARKET_MOBULA_CHAIN_ID`
- `EXCHANGE_RATES_MOBULA_API_KEY` -> `MARKET_MOBULA_API_KEY`
- `EXCHANGE_RATES_MOBULA_COIN_ID` -> `MARKET_MOBULA_COIN_ID`
- `EXCHANGE_RATES_MOBULA_SECONDARY_COIN_ID` -> `MARKET_MOBULA_SECONDARY_COIN_ID`


PR to docs will be done when env changes will be reviewed and confirmed


<details><summary>Configuration how to (internal link https://www.notion.so/blockscout/Exchange-rates-configuration-19e3d73641f8804396fed75cbb8eb7bc):</summary>
<p>

To disable all fetchers **and** any market data displaying: **`DISABLE_MARKET: true` 

Attention for splitted API/Indexer setup: Token and History fetching runs in indexer mode, Coin fetching runs in API mode**

## Coin exchange rate

To enable coin exchange rate: set any of
**`MARKET_COINGECKO_COIN_ID`**

- how to set
    
    Use [`https://api.coingecko.com/api/v3/coins/list`](https://api.coingecko.com/api/v3/coins/list) endpoint, find the coin you need, for instance, if you need Optimism, you need this object in the list from response
    
    ```
      {
        "id": "optimism",
        "symbol": "op",
        "name": "Optimism"
      },
    ```
    
    So, you need to set **`MARKET_COINGECKO_COIN_ID=optimism`**
    

**`MARKET_COINMARKETCAP_COIN_ID`** 

- how to set
    
    Use [`[https://pro-api.coinmarketcap.com/v1/cryptocurrency/map](https://pro-api.coinmarketcap.com/v1/cryptocurrency/map%5C%5C?CMC_PRO_API_KEY%5C%5C=c7f3a574-c85)`](https://pro-api.coinmarketcap.com/v1/cryptocurrency/map%5C%5C?CMC_PRO_API_KEY%5C%5C=c7f3a574-c85) endpoint (optionally you can use `symbol` query parameter), find the coin you need, for instance, if you need Optimism, you need this object in the list from response
    
    ```jsx
    {
        "id": 11840,
        "rank": 57,
        "name": "Optimism",
        "symbol": "OP",
        "slug": "optimism-ethereum",
        "is_active": 1,
        "status": 1,
        "first_historical_data": "2022-05-31T16:30:00.000Z",
        "last_historical_data": "2025-03-25T09:10:00.000Z",
        "platform": {
          "id": 42,
          "name": "Optimism",
          "symbol": "OP",
          "slug": "optimism-ethereum",
          "token_address": "0x4200000000000000000000000000000000000042"
        }
      }
    ```
    
    So, you need to set **`MARKET_COINMARKETCAP_COIN_ID=11840`**
    

**`MARKET_CRYPTORANK_COIN_ID`**

- how to set
    
    Use [`https://api.cryptorank.io/v1/currencies?symbols=OP`](https://api.cryptorank.io/v1/currencies?symbols=OP) endpoint (note that symbols value is case sensitive), find the coin you need, for instance, if you need Optimism, you need this object in the list from response
    
    ```jsx
    {
                "id": 174889,
                "rank": 62,
                "slug": "optimism",
                "name": "Optimism",
                "symbol": "OP",
                "category": "Chain",
                "type": "token",
                "volume24hBase": 42513205.1808,
                "circulatingSupply": 1620705846,
                "totalSupply": 4294967296,
                "maxSupply": 4294967296,
                "values": {
                    "USD": {
                        "price": 0.909209275302,
                        "volume24h": 38653400,
                        "high24h": 0.913696529092,
                        "low24h": 0.886532266507,
                        "marketCap": 1473560787.719375,
                        "percentChange24h": 1.081,
                        "percentChange7d": 2.8909,
                        "percentChange30d": -22.136,
                        "percentChange3m": -53.3746,
                        "percentChange6m": -47.5585
                    }
                },
                "lastUpdated": "2025-03-25T09:40:46.136Z",
                "tokens": [
                    {
                        "tokenAddress": "0x4200000000000000000000000000000000000042",
                        "platform": {
                            "id": 84,
                            "name": "Optimism",
                            "slug": "optimism"
                        }
                    }
                ]
            }
    ```
    
    So, you need to set **`MARKET_CRYPTORANK_COIN_ID=174889`**
    

**`MARKET_MOBULA_COIN_ID`** 

- how to set
    
    Use [`https://api.mobula.io/api/1/all`](https://api.mobula.io/api/1/all) endpoint, find the coin you need, for instance, if you need Optimism, you need this object in the list from response
    
    ```jsx
    {
      "id": 100552373,
      "name": "Optimism",
      "symbol": "OP"
    }
    ```
    
    So, you need to set **`MARKET_MOBULA_COIN_ID=100552373`**
    

To enable secondary coin exchange rate: set any of
**`MARKET_COINGECKO_SECONDARY_COIN_ID`
`MARKET_COINMARKETCAP_SECONDARY_COIN_ID`
`MARKET_CRYPTORANK_SECONDARY_COIN_ID`
`MARKET_MOBULA_SECONDARY_COIN_ID`**

Same instructions on how to set as for regular coin above

To override selected source use:
**`MARKET_NATIVE_COIN_SOURCE`** for coin
**`MARKET_SECONDARY_COIN_SOURCE`** for secondary coin

Possible values: **`coin_gecko`, `coin_market_cap`, `crypto_rank`, `mobula`**
When it could be needed: you have set **`MARKET_COINGECKO_COIN_ID`** and **`MARKET_COINMARKETCAP_COIN_ID`**, you want coinmarketcap for current price fecting and coingecko for history fetching, however, by default, coingecko will be used for both, to achieve what you want set **`MARKET_NATIVE_COIN_SOURCE: coin_market_cap`**

**Default source: CoinGecko**

To disable fetcher: 
**`MARKET_COIN_FETCHER_ENABLED: false`** enabled by default

## Token exchange rate

To enable token exchange rates: set any of
**`MARKET_COINGECKO_PLATFORM_ID`**

- how to set
    
    Use [`https://api.coingecko.com/api/v3/asset_platforms`](https://api.coingecko.com/api/v3/asset_platforms) endpoint, find the object with the network that you need, for instance, Optimism:
    
    ```
    {
        "id": "optimistic-ethereum",
        "chain_identifier": 10,
        "name": "Optimism",
        "shortname": "Optimism",
        "native_coin_id": "ethereum",
        "image": {
          "thumb": "https://coin-images.coingecko.com/asset_platforms/images/41/thumb/optimism.png?1706606778",
          "small": "https://coin-images.coingecko.com/asset_platforms/images/41/small/optimism.png?1706606778",
          "large": "https://coin-images.coingecko.com/asset_platforms/images/41/large/optimism.png?1706606778"
        }
      }
    ```
    
    Set **`MARKET_COINGECKO_PLATFORM_ID=optimistic-ethereum`**
    

**`MARKET_CRYPTORANK_PLATFORM_ID`** 

- how to set
    
    You can look for the network you need under `platform` field in `tokens` list in  [`https://api.cryptorank.io/v1/currencies?symbols=OP`](https://api.cryptorank.io/v1/currencies?symbols=OP) response
    
    ```json
    {
        "id": 174889,
        "rank": 62,
        "slug": "optimism",
        "name": "Optimism",
        "symbol": "OP",
        "category": "Chain",
        "type": "token",
        "volume24hBase": 42992154.9665,
        "circulatingSupply": 1620705846,
        "totalSupply": 4294967296,
        "maxSupply": 4294967296,
        "values": {
            "USD": {
                "price": 0.911420545028,
                "volume24h": 39183933,
                "high24h": 0.913696529092,
                "low24h": 0.886532266507,
                "marketCap": 1477144605.4913857,
                "percentChange24h": 1.5092,
                "percentChange7d": 3.4349,
                "percentChange30d": -21.7551,
                "percentChange3m": -53.2612,
                "percentChange6m": -47.4309
            }
        },
        "lastUpdated": "2025-03-25T09:47:46.823Z",
        "tokens": [
            {
                "tokenAddress": "0x4200000000000000000000000000000000000042",
                "platform": {
                    "id": 84,
                    "name": "Optimism",
                    "slug": "optimism"
                }
            }
        ]
    }
    ```
    
    Set **`MARKET_CRYPTORANK_PLATFORM_ID=84`**
    

**`MARKET_MOBULA_CHAIN_ID`**

- how to set
    
    Use [`https://api.mobula.io/api/1/blockchains`](https://api.mobula.io/api/1/blockchains) endpoint, find the object with the network that you need, for instance, Optimism:
    
    ```json
      {
          "name": "Optimistic",
          "rpcs": ["https://mainnet.optimism.io"],
          "explorer": "https://optimistic.etherscan.io",
          "color": "#FF1D26",
          "chainId": "10",
          "evmChainId": 10,
          "multicall_contract": "0x142e2feac30d7fc3b61f9ee85fccad8e560154cc",
          "uniswapV3Factory": ["0x1f98431c8ad98523631ae4a59f267346ea31f984"],
          "eth": {
            "name": "Ethereum",
            "symbol": "ETH",
            "address": "0x4200000000000000000000000000000000000006",
            "type": "eth",
            "decimals": 18,
            "logo": "https://metacore.mobula.io/60989b438a473ed77703f7bcd3530fc15542996a6607ae831de8c5e3e10f688d.png",
            "id": 100004304
          },
          "stable": {
            "name": "USD Coin",
            "symbol": "USDC",
            "address": "0x7f5c764cbc14f9669b88837ca1490cca17c31607",
            "type": "stable",
            "decimals": 6,
            "logo": "https://s2.coinmarketcap.com/static/img/coins/64x64/3408.png",
            "blockchain": "Optimistic",
            "blockchains": [
              "Ethereum",
              "Polygon",
              "BNB Smart Chain (BEP20)",
              "Fantom",
              "Avalanche C-Chain",
              "SmartBCH",
              "Cronos",
              "DFK Subnet",
              "Celo",
              "XDAI",
              "Klaytn",
              "Aurora",
              "HECO",
              "Harmony",
              "Boba",
              "OKEX",
              "Moonriver",
              "Canto",
              "BitTorrent Chain",
              "Oasis",
              "Velas",
              "Arbitrum",
              "Optimistic",
              "Kucoin",
              "Moonbeam",
              "Mantle",
              "ZkSync",
              "Linea",
              "Metis",
              "Arbitrum Nova",
              "Scroll",
              "Polygon zkEVM",
              "Manta"
            ],
            "contracts": [
              "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
              "0x2791bca1f2de4661ed88a30c99a7a9449aa84174",
              "0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d",
              "0x04068da6c83afcfa0e13ba15a6696662335d5b75",
              "0xb97ef9ef8734c71904d8002f8b6bc66dd9c48a6e",
              "0xbcbd9990dcec6a64741ea27bec0ca8ff6b91bc26",
              "0xc21223249ca28397b4b6541dffaecc539bff0c59",
              "0x3ad9dfe640e1a9cc1d9b0948620820d975c3803a",
              "0x765de816845861e75a25fca122bb6898b8b1282a",
              "0xddafbb505ad214d7b80b1f830fccc89b60fb7a83",
              "0x754288077d0ff82af7a5317c7cb8c444d421d103",
              "0xb12bfca5a55806aaf64e99521918a4bf0fc40802",
              "0x9362bbef4b8313a8aa9f0c9808b80577aa26b73b",
              "0x985458e523db3d53125813ed68c274899e9dfab4",
              "0x66a2a913e447d6b4bf33efbec43aaef87890fbbc",
              "0xc946daf81b08146b1c7a8da2a851ddf2b3eaaf85",
              "0xe3f5a90f9cb311505cd691a46596599aa1a0ad7d",
              "0x80b5a32e4f032b2a058b4f29ec95eefeeb87adcd",
              "0x935faa2fcec6ab81265b301a30467bbc804b43d3",
              "0xe8a638b3b7565ee7c5eb9755e58552afc87b94dd",
              "0xe2c120f188ebd5389f71cf4d9c16d05b62a58993",
              "0xff970a61a04b1ca14834a43f5de4533ebddb5cc8",
              "0x7f5c764cbc14f9669b88837ca1490cca17c31607",
              "0x980a5afef3d17ad98635f6c5aebcbaeded3c3430",
              "0x818ec0a7fe18ff94269904fced6ae3dae6d6dc0b",
              "0x09bc4e0d864854c6afb6eb9a9cdf58ac190d0df9",
              "0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4",
              "0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f",
              "0xea32a96608495e54156ae48931a7c20f0dcc1a21",
              "0x750ba8b76187092b0d1e87e28daaf484d1b5273b",
              "0x06efdbff2a14a7c8e15944d1f4a48f9f95f663a4",
              "0xa8ce8aee21bc2a48a5ef670afcc9274c7bbbc035",
              "0xb73603c5d87fa094b7314c74ace2e64d165016fb"
            ]
          },
          "routers": [
            {
              "address": "0x9c12939390052919af3155f41bf4160fd3666a6f",
              "name": "Velodrome Finnance",
              "factory": "0x25cbddb98b35ab1ff77413456b31ec81a6b6b746",
              "fee": 9970
            }
          ],
          "tokens": [
            {
              "address": "0x4200000000000000000000000000000000000006",
              "name": "Ethereum",
              "symbol": "ETH",
              "decimals": 18,
              "type": "eth"
            },
            {
              "address": "0x7f5c764cbc14f9669b88837ca1490cca17c31607",
              "name": "USD Coin",
              "symbol": "USDC",
              "decimals": 6,
              "type": "stable"
            },
            {
              "address": "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
              "name": "Tether",
              "symbol": "USDT",
              "decimals": 6,
              "type": "stable"
            },
            {
              "address": "0x4200000000000000000000000000000000000042",
              "name": "Optimism Token",
              "symbol": "OPT",
              "decimals": 18,
              "type": "stable"
            }
          ],
          "supportedProtocols": ["Li.Fi", "OneInch"],
          "logo": "https://assets.coingecko.com/coins/images/25244/large/Optimism.png?1660904599",
          "coingeckoChain": "optimistic-ethereum",
          "dexscreenerChain": "optimism",
          "isLayer2": true,
          "coverage": ["metadata", "market", "explorer"]
        }
    ```
    
    Set either **`MARKET_MOBULA_CHAIN_ID=Optimism`** or **`MARKET_MOBULA_CHAIN_ID=10`** (`name` or `chainId` should both work)
    

Source can be selected manually via **`MARKET_TOKENS_SOURCE`**
Possible values: **`coin_gecko`, `crypto_rank`, `mobula`**

**Default source: CoinGecko**
To disable fetcher:
**`MARKET_TOKENS_FETCHER_ENABLED: false`** enabled by default

## History fetching

To enable price history fetching: set any of
**`MARKET_COINGECKO_COIN_ID`
`MARKET_COINMARKETCAP_COIN_ID` 
`MARKET_CRYPTOCOMPARE_COIN_SYMBOL`** or **`COIN`**

- how to set
    
    CryptoCompare does not use ids, so here simpy put your coin symbol, for instance, `OP`
    

**`MARKET_CRYPTORANK_COIN_ID`
`MARKET_MOBULA_COIN_ID`**

Instruction for other provides can be found in [[Coin exchange rate](https://www.notion.so/Coin-exchange-rate-19e3d73641f880ccba8fc64786736127?pvs=21)](https://www.notion.so/Coin-exchange-rate-19e3d73641f880ccba8fc64786736127?pvs=21) 

To override: **`MARKET_NATIVE_COIN_HISTORY_SOURCE`**

**Default source: CryptoCompare**

To enable secondary coin price fetching set any of

**`MARKET_COINGECKO_SECONDARY_COIN_ID`
`MARKET_COINMARKETCAP_SECONDARY_COIN_ID` 
`MARKET_CRYPTOCOMPARE_SECONDARY_COIN_SYMBOL`**
Same as above
**`MARKET_CRYPTORANK_SECONDARY_COIN_ID`
`MARKET_MOBULA_SECONDARY_COIN_ID`**

To override: **`MARKET_SECONDARY_COIN_HISTORY_SOURCE`**

**Default source: CryptoCompare**

To enable market cap fetching set any of
**`MARKET_COINGECKO_COIN_ID`
`MARKET_COINMARKETCAP_COIN_ID`**

To override: **`MARKET_MARKET_CAP_HISTORY_SOURCE`**

**Default source: CoinGecko**

To enable tvl history fetching set
**`MARKET_DEFILLAMA_COIN_ID`**

- how to set
    
    Use [`https://api.llama.fi/v2/chains`](https://api.llama.fi/v2/chains) endpoint, find the chain you need in the response, for instance, Optimism is:
    
    ```
    {
      "gecko_id": "optimism",
      "tvl": 483265534.4437918,
      "tokenSymbol": "OP",
      "cmcId": "11840",
      "name": "OP Mainnet",
      "chainId": 10
    }
    ```
    
    Set **`MARKET_DEFILLAMA_COIN_ID=OP Mainnet`** (field `name` from the object)
    

To override: **`MARKET_TVL_HISTORY_SOURCE`**

**Default source: DefiLlama**

</p>
</details> 

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/setup/env-variables/README.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/setup/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/setup/env-variables). https://github.com/blockscout/docs/pull/399
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced unified market data fetching components that retrieve coin, token, and historical market information.
  
- **Refactor**
  - Replaced legacy USD-specific financial metrics with generalized fiat currency values across API responses, UI templates, CSV exports, and notifications.
  - Restructured configuration settings to use new market data sources instead of old exchange rate modules.

- **Chores**
  - Cleaned up redundant modules and updated database schema (e.g., tokens icon URL) to support the new market data framework.
  
- **Tests**
  - Updated test suites to align with the new market data logic and configuration changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sl1depengwyn commented:
@coderabbitai review

vbaranov commented on apps/explorer/lib/explorer/chain/import/runner/tokens.ex:
```diff
@@ -241,10 +252,45 @@ defmodule Explorer.Chain.Import.Runner.Tokens do
   end
 
-  defp should_update?(_new_token, nil), do: true
+  def market_data_on_conflict do
```
Maybe we could move this on confilct function to Explorer.Market.Fetcher.Token since it is used only there and turn it to private?

sl1depengwyn replied:
I think here applies the same as in the previous comment

vbaranov reviewed (commented):
I have a concern regarding default values for native coin price and market cap. Before the PR default values are:
price -> CryptoCompare
market cap -> CoinGecko

In many setups we rely on these default values. If I am not mistaken at least default source for price fetching is changed in this PR. Can we keep current default values in order to simplify update procedure?

vbaranov commented:
@sl1depengwyn also, please rebase the PR in order to resolve merging conflicts and fix explorer tests.

sl1depengwyn commented on apps/explorer/priv/repo/migrations/20250217221717_change_tokens_icon_url_type.exs:
```diff
@@ -0,0 +1,9 @@
+
+  def change do
+    alter table(:tokens) do
+      modify(:icon_url, :text, null: true)
```
Yes, I faced such case with Mobula, here is an example: https://coin-images.coingecko.com/coins/images/38457/large/%E0%B8%94%E0%B8%B5%E0%B9%84%E0%B8%8B%E0%B8%99%E0%B9%8C%E0%B8%97%E0%B8%B5%E0%B9%88%E0%B8%A2%E0%B8%B1%E0%B8%87%E0%B9%84%E0%B8%A1%E0%B9%88%E0%B9%84%E0%B8%94%E0%B9%89%E0%B8%95%E0%B8%B1%E0%B9%89%E0%B8%87%E0%B8%8A%E0%B8%B7%E0%B9%88%E0%B8%AD_%282%29.png?1717575131

I think DoS here is not probable as we use more or less trusted sources, do you think its better to just increase the limit to 1024, for instance?

sl1depengwyn commented on apps/explorer/lib/explorer/chain/import/runner/tokens.ex:
```diff
@@ -241,10 +252,45 @@ defmodule Explorer.Chain.Import.Runner.Tokens do
+    )
+  end
+
+  def market_data_fields_to_update do
```
I think it's better to keep it in `Explorer.Chain.Import.Runner.Tokens`, since its functionality tightly bound to this module. For instance, if one would want to update current on_conflict clause with new field chances that `market_data_fields_to_update` would be forgotten since it placed not in the same module with `default_on_conflict`

sl1depengwyn commented on apps/explorer/lib/explorer/market/source.ex:
```diff
@@ -0,0 +1,369 @@
+  @spec native_coin_price_history_source() :: module
+  def native_coin_price_history_source do
+    config(:native_coin_history_source) ||
+      Enum.find([CryptoCompare | @sources], fn source ->
```
No, since it other places `CoinGecko` should be the default source

sl1depengwyn commented on apps/explorer/lib/explorer/market/fetcher/token.ex:
```diff
@@ -0,0 +1,120 @@
+    {:noreply, state}
+  end
+
+  defp update_tokens(token_params) do
+    Chain.import(%{
+      tokens: %{
+        params: token_params,
+        on_conflict: Tokens.market_data_on_conflict(),
+        fields_to_update: Tokens.market_data_fields_to_update()
+      }
+    })
+  end
+
```
It's handled in `Chain.import` function

sl1depengwyn commented on apps/explorer/lib/explorer/market/fetcher/coin.ex:
```diff
@@ -0,0 +1,243 @@
+    end
+  end
+
+  @doc """
+  Lists exchange rates for the tracked tickers.
+  """
+  @spec list :: [Token.t()] | nil
+  def list do
+    if config(:enabled) do
+      list_from_store(config(:store))
+    end
+  end
```
Actually it wasn't good catch

vbaranov reviewed (commented):
Please rebase for the sake of resolving merge conflicts. I didn't find the docs PR yet. It is not urgent since you added backward compatibility, but would be good to have it merged with public release publishing.

sl1depengwyn commented:
> Please rebase for the sake of resolving merge conflicts. I didn't find the docs PR yet. It is not urgent since you added backward compatibility, but would be good to have it merged with public release publishing.

I've rebased the PR and updated the docs https://github.com/blockscout/docs/pull/399

diaztorresangeladrian61 commented:
`apps/block_scout_web/test/block_scout_web/channels/v2/exchange_rate_channel_test.exs`
